pytorch_lightning/utilities/auto_restart.py

```diff
@@ -116,13 +125,114 @@ def _compute_current_iteration(self, num_batches_processed: Optional[int] = None
 
         return current_iteration
 
-    def _load_cached_state(self):
-        if self._cached_state_dict is None or self.worker_id not in self._cached_state_dict:
-            return
-        self._current_iteration = self._cached_state_dict[self.worker_id]["current_iteration"]
-        # delete cached state, prevent reloading every time iter() is called
+    def _load_non_random_state(self, state_dict: Dict[int, Dict[str, Any]]) -> None:
+        self._current_iteration = state_dict[self.worker_id]["current_iteration"]
+
+
+@dataclass(frozen=True, unsafe_hash=True)
+class IteratorState:
+    dataset_state: Dict[int, Any] = field(default_factory=dict)
+    sampler_state: Dict[int, Any] = field(default_factory=dict)
+    worker_id: int = 0
+    num_workers: int = 0
+    num_batches_fetched: int = 0
+    name: Optional[str] = None
+
+    @classmethod
+    def load_state_dict(cls, state_dict) -> "IteratorState":
+        return cls(**state_dict)
+
+
+@dataclass
+class CollectionIteratorState:
+    """This class is used to hold the current iterator state and lives on the iterator."""
+
+    state: Union[Dict[Union[int, str], Union[Dict[str, IteratorState], IteratorState]]] = field(default_factory=dict)
+    lastest_worker_id: int = 0
+    represent_map_dataset: Optional[bool] = None
+
+    def update(self, iter_name: Optional[str], new_state: IteratorState) -> None:
+        self.represent_map_dataset = iter_name is None
+        if self.represent_map_dataset:
+            state = self.state
+        else:
+            if iter_name not in self.state:
+                self.state[iter_name] = {}
+            state = self.state[iter_name]
+
+        lastest_worker_id = new_state.worker_id
+        state[lastest_worker_id] = new_state
+        self.lastest_worker_id = lastest_worker_id
+
+    @property
+    def sampler_states(self) -> Dict[int, Any]:
+        return {0: self.state[k].sampler_state[0] for k in self.state.keys()}
```

Comment:
if all is with key 0, it means only the last is kept?
```suggestion
        return {k: self.state[k].sampler_state[0] for k in self.state.keys()}
```

Reply:
these methods are currently unused, I removed them. 
I will double check and address your comments in the follow up PR! Thanks